pkg/kubelet/allocation/allocation_manager.go

```diff
@@ -61,18 +64,56 @@ type Manager interface {
 
 	// RemoveOrphanedPods removes the stored state for any pods not included in the set of remaining pods.
 	RemoveOrphanedPods(remainingPods sets.Set[types.UID])
+
+	// Run starts the allocation manager. This is currently only used to handle periodic retry of
+	// pending resizes.
+	Run()
+
+	// PushPendingResize queues a pod with a pending resize request for later reevaluation.
+	// If a change in the resize queue is detected, pending resizes will be retried.
+	PushPendingResize(pod *v1.Pod)
+
+	// RetryPendingResizes signals the allocation manager to retry all pending resize requests.
+	RetryPendingResizes()
+
+	// HandlePodResourcesResize determines if a pod resize is feasible. If so,
+	// it updates the allocation and sets the pod resize conditions accordingly.
+	HandlePodResourcesResize(pod *v1.Pod) error
 }
 
 type manager struct {
 	allocated state.State
 	actuated  state.State
+
+	canResizePod   func(pod *v1.Pod) (bool, string, string)
+	triggerPodSync func(pod *v1.Pod)
+
+	statusManager       status.Manager
+	skipPeriodicRetries bool
+	pendingResizeChan   chan struct{}
+
+	pendingResizesLock     sync.RWMutex
+	podsWithPendingResizes []*v1.Pod
+
+	podResizeMutex *sync.Mutex
 }
 
-func NewManager(checkpointDirectory string) Manager {
-	return &manager{
-		allocated: newStateImpl(checkpointDirectory, allocatedPodsStateFile),
-		actuated:  newStateImpl(checkpointDirectory, actuatedPodsStateFile),
-	}
+func NewManager(
+	checkpointDirectory string,
+	statusManager status.Manager,
+	podResizeMutex *sync.Mutex,
```

Comment:
Passing mutexes is a code smell. The fact that this is necessary makes me think we might have the wrong abstraction here.

The other place this is needed is in `HandlePodAdditions`, but that's because adding a pod is actually an allocation action. I'm thinking that rather than sharing a mutex, we should just move the pod addition allocation logic into the allocation manager. In other words, add a method like `AllocatePod` (or AddPod, if you prefer), which does:
1. lock mutex
2. run admission check (or just the fit check, if we want to split that out from admission)
3. update the pod allocation, or return a failure if the admission check failed

We might want to discuss this approach more offline.

Reply:
thanks, the offline discussion helped me understand what we are going for

moving the admission check into the allocation manager turned into its own little project so I broke it off into another PR: https://github.com/kubernetes/kubernetes/pull/131801

If we take https://github.com/kubernetes/kubernetes/pull/131801, I'll rebase this one on it